### PR TITLE
Kmcp compute

### LINKED ISSUE #3117: new module: KMCP/COMPUTE
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by sofstam**
## PR checklist

Closes #3117  

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/kmcp/compute/main.nf [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    conda "bioconda::kmcp=0.9.1"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/kmcp:0.9.1--h9ee0642_0':
+        'quay.io/biocontainers/kmcp:0.9.1--h9ee0642_0' }"
+
+    input:
+    tuple val(meta), path(sequences)
+
+    output:
+    tuple val(meta), path("tmp/*") , emit: tmp
```
Comment (jfy133) [inline modules/nf-core/kmcp/compute/main.nf:14]: Is `tmp` really the name of the output?

File: modules/nf-core/kmcp/compute/main.nf [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def input = sequences.isDirectory()? "--in-dir ${sequences}" : "${sequences}"
+    """
+    kmcp \\
+        compute \\
+        $args \\
+        --threads $task.cpus \\
+        --out-dir 'tmp/' \\
```
Comment (jfy133) [inline modules/nf-core/kmcp/compute/main.nf:29]: Why `tmp/` and not `${prefix}/`?

File: modules/nf-core/kmcp/compute/main.nf (hunk 1 above)
Comment (jfy133) [reply modules/nf-core/kmcp/compute/main.nf:14]: Also I would recommend also exporting `_info.txt` which is in the directory as a separate thing too (as wella sin the directory)

File: modules/nf-core/kmcp/compute/meta.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - sequences:
+      type: file
+      description: fasta file
```
Comment (jfy133) [inline modules/nf-core/kmcp/compute/meta.yml:22]: Or directory, correct?

File: modules/nf-core/kmcp/compute/meta.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - tmp:
```
Comment (jfy133) [inline modules/nf-core/kmcp/compute/meta.yml:35]: See comment above. Maybe mention `.unik` are the kmer sketches.

File: tests/modules/nf-core/kmcp/compute/test.yml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+- name: kmcp compute test_kmcp_compute
+  command: nextflow run ./tests/modules/nf-core/kmcp/compute -entry test_kmcp_compute -c ./tests/config/nextflow.config -c ./tests/modules/nf-core/kmcp/compute/nextflow.config
+  tags:
+    - kmcp/compute
+    - kmcp
+  files:
+    - path: output/kmcp/tmp/_info.txt
```
Comment (jfy133) [inline tests/modules/nf-core/kmcp/compute/test.yml:7]: Is there no nice string in there *e.g. 'completed') or something to indicate it ran correctly?

**Review by jfy133 — CHANGES_REQUESTED** [review]
My main quibble is the outoput directory name :)

File: modules/nf-core/kmcp/compute/main.nf (hunk 2 above)
Comment (sofstam) [reply modules/nf-core/kmcp/compute/main.nf:29]: For some reason thought that tmp was standard name for the output directory.

File: modules/nf-core/kmcp/compute/main.nf (hunk 1 above)
Comment (jfy133) [reply modules/nf-core/kmcp/compute/main.nf:14]: https://nf-co.re/docs/contributing/modules#naming-conventions point 5 :)
Then in the tesdt you update prefix in the nextflow.config

File: modules/nf-core/kmcp/compute/main.nf (hunk 2 above)
Comment (jfy133) [reply modules/nf-core/kmcp/compute/main.nf:29]: Fair enough - I guess change as above :)

File: modules/nf-core/kmcp/compute/meta.yml (hunk 3 above)
Comment (sofstam) [reply modules/nf-core/kmcp/compute/meta.yml:22]: Yes!

File: modules/nf-core/kmcp/compute/meta.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
+name: "kmcp_compute"
+description: Generate k-mers (sketches) from FASTA/Q sequences
+keywords:
+  - metagenomics
+  - classify
+  - taxonomic profiling
+  - fastq
+  - sequences
```
Comment (jfy133) [inline modules/nf-core/kmcp/compute/meta.yml:8]:
```suggestion
  - sequences
  - kmers
```

File: modules/nf-core/kmcp/compute/meta.yml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - sequences:
+      type:
+        - file
+        - directory
+      description: fasta file
```
Comment (jfy133) [inline modules/nf-core/kmcp/compute/meta.yml:28]:
```suggestion
      description: fasta file, or a directory containing FASTA files
```

File: modules/nf-core/kmcp/compute/meta.yml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      pattern: "*.{fa,fa.gz,fasta,fasta.gz,fna,fna.gz}"
```
Comment (jfy133) [inline modules/nf-core/kmcp/compute/meta.yml:29]:
```suggestion
      pattern: "**/*.{fa,fa.gz,fasta,fasta.gz,fna,fna.gz}"
```

File: modules/nf-core/kmcp/compute/main.nf [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/kmcp:0.9.1--h9ee0642_0':
+        'quay.io/biocontainers/kmcp:0.9.1--h9ee0642_0' }"
+
+    input:
+    tuple val(meta), path(sequences)
+
+    output:
+    tuple val(meta), path("${prefix}/*")           , emit: tmp
+    tuple val(meta), path("${prefix}/_info.txt")           , emit: info
```
Comment (jfy133) [inline modules/nf-core/kmcp/compute/main.nf:15]:
```suggestion
    tuple val(meta), path("${prefix}/*")           , emit: outdir
    tuple val(meta), path("${prefix}/_info.txt")   , emit: info
```

File: modules/nf-core/kmcp/compute/meta.yml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - tmp:
```
Comment (jfy133) [inline modules/nf-core/kmcp/compute/meta.yml:41]:
```suggestion
  - outdir:
```

**Review by jfy133 — APPROVED** [review]
Accept the following changes, wait for tests to pass and then you're good to go :)

**Pull request merged**